pkg/validation/component.go

```diff
@@ -295,3 +306,104 @@ func (c *component) reportEvent(ctx context.Context, apiCall string, values ...s
 		internal.LogUnrecordedEvent(ctx, c.logger, apiCall, errEvent.Error(), values...)
 	}
 }
+
+type validationContext struct {
+	ctx         context.Context
+	component   *component
+	accessToken *string
+	realmName   string
+	userID      string
+	kcUser      *kc.UserRepresentation
+	dbUser      *dto.DBUser
+}
+
+func (v *validationContext) getAccessToken() (string, error) {
+	if v.accessToken == nil {
+		if accessToken, err := v.component.tokenProvider.ProvideToken(v.ctx); err == nil {
+			v.accessToken = &accessToken
+		} else {
+			v.component.logger.Warn(v.ctx, "msg", "Can't get access token", "err", err.Error(), "realm", v.realmName, "user", v.userID)
+			return "", err
+		}
+	}
+	return *v.accessToken, nil
+}
+
+func (v *validationContext) getKeycloakUser() (*kc.UserRepresentation, error) {
+	if v.kcUser == nil {
+		var accessToken, err = v.getAccessToken()
+		if err != nil {
+			return nil, err
+		}
+		if kcUser, err := v.component.keycloakClient.GetUser(accessToken, v.realmName, v.userID); err == nil {
+			v.kcUser = &kcUser
+		} else {
+			v.component.logger.Warn(v.ctx, "msg", "Can't get user from Keycloak", "err", err.Error(), "realm", v.realmName, "user", v.userID)
+			return nil, err
+		}
+	}
+	return v.kcUser, nil
+}
+
+func (v *validationContext) updateKeycloakUser() error {
+	accessToken, err := v.getAccessToken()
+	if err != nil {
+		return err
+	}
+
+	err = v.component.keycloakClient.UpdateUser(accessToken, v.realmName, v.userID, *v.kcUser)
+	if err != nil {
+		v.component.logger.Warn(v.ctx, "msg", "updateKeycloakUser: can't update user in KC", "err", err.Error(), "realmName", v.realmName, "userID", v.userID)
+		return errorhandler.CreateInternalServerError("keycloak")
+	}
+	return nil
+}
+
+func (v *validationContext) getDbUser() (*dto.DBUser, error) {
+	if v.dbUser == nil {
+		if dbUser, err := v.component.usersDBModule.GetUserDetails(v.ctx, v.realmName, v.userID); err == nil {
+			v.dbUser = &dbUser
+		} else {
+			v.component.logger.Warn(v.ctx, "msg", "Can't get user from database", "err", err.Error(), "realm", v.realmName, "user", v.userID)
+			return nil, err
+		}
+	}
+	return v.dbUser, nil
+}
+
+func (v *validationContext) getUserWithAccreditations() (*kc.UserRepresentation, error) {
```

Comment:
I don't understnad why we have to eprform a lot more stuff in validation compare to kyc. 
Maybe we could move all this logic linked to the user archiving into module layer ?

Reply:
KYC only stores a kyc/UserRepresentation : all input fields are mandatory
Validation stores a validation/UserRepresentation merged with information from Keycloak and from database then add checks... input fields are not mandatory. A simple way to do this would be to keep the processings as their are already coded then just go and get information in Keycloak and database but it would be much less performant.
I'll try to do it a different way.